sklearn/base.py

```diff
@@ -133,7 +142,121 @@ def _pprint(params, offset=0, printer=repr):
     return lines
 
 
-class BaseEstimator:
+class _PropsRequest:
+    def get_props_request(self):
+        """Get requested data properties.
+
+        Returns
+        -------
+        props : dict of list of strings, or dict of dict of {str: str}, or None
+            The key to the top level dict is the method for which the prop is
+            used. Under each key, there is a dict list of required properties,
+            or a dict of mapping of the form
+            ``{method_parameter: provided_parameter}``, or None.
+        """
+        try:
+            return self._props_request
+        except AttributeError:
+            return {}
+
+    def set_props_request(self, props):
+        """Set required data properties.
+
+        Parameters
+        ----------
+        props : dict of list of strings, or dict of dict of {str: str}, or None
+            The key to the top level dict is the method for which the prop is
+            used. Under each key, there is a dict list of required properties,
+            or a dict of mapping of the form
+            ``{method_parameter: provided_parameter}``.
+            ``None`` empties the required props.
+
+        Returns
+        -------
+        self : estimator
+            returns self.
+        """
+        if props is None:
+            try:
+                del self._props_request
+            except AttributeError:
+                pass
+            return self
+
+        if not isinstance(props, dict):
+            raise ValueError("`props` should be a dictionary")
+
+        for method, value in props.items():
+            if isinstance(value, str):
+                props[method] = [value]
+
+        if not hasattr(self, '_props_request'):
+            self._props_request = {}
+
+        for method, m_props in props.items():
+            if method not in self._props_request:
+                self._props_request[method] = []
+
+            if isinstance(m_props, dict):
+                if isinstance(self._props_request[method], list):
```

Comment:
If a dict is the more general form of props request, should we not just always store it as a dict?